### PR TITLE
Not retaining thread in concurrent hash join threadpool

### PR DESCRIPTION
**Pull request by canhld94**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Not retaining thread in concurrent hash join threadpool to avoid query excessively spawn threads 

Motivation: last week our clusters suddenly became unstable, lots of queries stuck because they cannot allocate new threads. Tracing the trace log led us to a user that use `parallel_hash` by default with some big JOINs. Each query from this user use more than 200 threads and ddos the whole cluster.
This PR try to fix the issue by limiting size of thread pool in `ConcurrentHashJoin` to respect `max_threads` setting.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by canhld94** [comment]
Hm... something wrong, when building `ConcurrentHashJoin` we've already passed `slot_ = max_threads`.

**Comment by canhld94** [comment]
It's more likely that: we have many `ConcurrentHashJoin` and each of them using a pool with size of `max_threads`. In this case, the proper fix is to have shared thread pool among all `ConcurrentHashJoin` instances.

**Comment by canhld94** [comment]
Or, we need to set `max_free_threads` in the thread pool inside `ConcurrentHashJoin` to 0 so that when they're done (the thread pool is used only in constructor and destructor), all threads are released.

This is exactly what we observed last week.

1. Some queries uses > 600 threads:

```
prata096 :) select thread_ids from system.processes limit 1\G

SELECT thread_ids
FROM system.processes
LIMIT 1

Query id: 518b44ae-794a-4244-b227-5fbb9665e083

Row 1:
──────
thread_ids: [3775969,3776186,3776092,3776187,3776283,3776123,3776046,3776200,3776178,3776221,3776213,3776244,3776206,3776241,3776194,3776205,3776184,3776193,3776201,3776207,3776220,3776196,3776169,3776048,3776044,3776004,3776049,3776091,3776087,3775958,3776055,3775974,3776082,3776080,3776050,3776077,3761200,3776038,3776056,3776094,3776225,3775962,3776059,3775970,3761085,3776228,3761105,3776248,3761095,3776238,3776175,3776192,3776216,3776234,3761092,3776235,3776188,3761093,3776236,3776217,3761233,3776034,3776039,3761190,3772596,3771799,3776069,3776035,3775983,3776093,3775999,3776076,3771765,3776132,3776073,3775941,3776085,3775917,3776057,3776040,3775954,3776068,3776054,3776041,3776071,3776053,3776090,3776051,3775985,3775984,3776058,3776083,3776079,3776084,3776066,3776047,3776036,3776088,3776074,3776043,3776064,3776065,3776062,3776006,3776190,3762295,3776095,3761194,3776089,3776033,3776081,3776060,3776061,3776067,3775931,3776009,3776072,3776045,3775995,3776086,3776042,3776134,3771767,3776153,3771786,3771839,3768651,3776104,3771737,3761235,3776157,3771790,3776106,3771739,3761256,3761212,3776215,3761129,3776111,3771744,3761121,3768664,3771840,3776247,3761161,3775937,3761248,3776171,3771843,3768655,3761482,3776121,3771754,3761234,3761266,3776119,3771752,3761247,3771811,3776097,3771730,3761191,3776203,3762308,3771836,3776140,3771773,3761110,3771841,3768653,3761224,3775987,3761298,3761141,3776146,3771779,3761236,3776159,3771792,3761188,3761134,3771772,3776107,3771740,3761291,3762285,3776237,3761094,3776099,3771732,3771838,3761119,3762286,3768662,3776114,3771747,3776195,3761109,3776202,3771835,3761195,3761208,3776162,3771795,3761100,3776243,3761157,3761120,3776125,3771758,3761166,3761165,3761264,3761218,3761207,3776212,3761126,3771845,3775928,3761239,3776105,3771738,3775919,3761230,3761183,3776147,3771780,3775957,3761268,3775949,3761260,3776245,3761159,3776164,3771797,3761209,3776100,3771733,3775981,3761292,3761177,3761214,3776204,3761118,3771837,3776154,3771787,3761259,3776124,3771757,3776126,3771759,3776117,3771750,3761112,3761245,3776260,3761174,3776223,3761137,3776142,3771775,3776231,3761088,3761168,3776197,3771842,3768654,3761216,3761127,3776208,3768665,3761179,3771761,3776116,3771749,3776191,3776110,3771743,3776102,3771735,3761145,3776150,3771783,3761210,3771802,3761083,3761280,3776141,3771774,3761136,3776177,3771810,3761172,3776250,3761107,3776112,3771745,3771764,3776156,3762261,3762314,3776275,3761132,3771770,3761203,3761600,3761206,3776211,3761125,3768656,3771844,3776151,3771784,3761226,3776145,3771778,3761244,3771768,3776167,3771800,3771771,3776130,3771763,3776098,3771731,3776109,3771742,3776101,3771734,3776182,3761096,3761193,3771766,3761151,3776189,3761103,3776108,3771741,3776158,3762263,3771791,3761232,3761170,3776199,3776127,3771760,3776103,3771736,3771801,3776249,3761163,3775908,3761219,3771798,3761225,3776152,3771785,3771753,3761115,3776170,3776227,3761084,3775967,3761278,3761229,3776148,3771781,3761086,3776229,3761143,3776113,3771746,3776210,3761124,3761173,3761270,3776118,3771751,3761241,3761180,3761182,3776122,3771755,3776161,3771794,3761099,3776242,3761156,3775942,3761253,3761288,3776096,3771729,3776143,3771776,3776230,3761144,3776149,3771782,3776115,3771748,3776136,3771769,3761215,3771756,3776155,3771788,3761204,3776144,3771777,3775944,3761255,3776172,3771796,3762268,3761164,3775940,3761251,3771793,3776129,3771762,3761262]

prata096 :) select length(thread_ids) from system.processes

SELECT length(thread_ids)
FROM system.processes

Query id: 819011bf-3b0b-4154-8250-b9c6297a3f53

    ┌─length(thread_ids)─┐
 1. │                681 │
 2. │                682 │
 3. │                730 │
 4. │                688 │
 5. │                520 │
 6. │                 26 │
 7. │                748 │
 8. │                727 │
 9. │                731 │
10. │                753 │
11. │                 26 │
12. │                  2 │
13. │                  2 │
14. │                497 │
15. │                 26 │
16. │                  2 │
17. │                 74 │
18. │                  2 │
19. │                  2 │
20. │                  1 │
21. │                  2 │
22. │                  2 │
23. │                  1 │
24. │                  1 │
    └────────────────────┘
```

2. Most of the thread are inactive:

```
SELECT
    count(),
    arrayStringConcat(arrayMap(x -> demangle(addressToSymbol(x)), trace), '\n') AS trace_symbols
FROM system.stack_trace
WHERE thread_id IN (
    SELECT arrayJoin(arrayConcat(thread_ids))
    FROM system.processes
    WHERE 1
)
GROUP BY trace_symbols
ORDER BY count() DESC
LIMIT 5
SETTINGS allow_introspection_functions = 1

Query id: b21e4bad-2e30-4f10-afe2-3c5412536161

Row 1:
──────
count():       1016
trace_symbols: DB::(anonymous namespace)::signalHandler(int, siginfo_t*, void*)



ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::worker(std::__1::__list_iterator<ThreadFromGlobalPoolImpl<false, true>, void*>)
void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*)
void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*)



Row 2:
──────
count():       530
trace_symbols: DB::(anonymous namespace)::signalHandler(int, siginfo_t*, void*)



void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*)



Row 3:
──────
count():       14
trace_symbols: DB::(anonymous namespace)::signalHandler(int, siginfo_t*, void*)



DB::ParallelFormattingOutputFormat::collectorThreadFunction(std::__1::shared_ptr<DB::ThreadGroup> const&)
void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true, true>::ThreadFromGlobalPoolImpl<DB::ParallelFormattingOutputFormat::ParallelFormattingOutputFormat(DB::ParallelFormattingOutputFormat::Params)::'lambda'()>(DB::ParallelFormattingOutputFormat::ParallelFormattingOutputFormat(DB::ParallelFormattingOutputFormat::Params)::'lambda'()&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*)
void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*)
```
3. Queries has many lock contention on acquiring threads on thread pool:

```
WITH
    (
        SELECT now() - 3600
    ) AS start_time,
    (
        SELECT now()
    ) AS end_time
SELECT
    arrayStringConcat(arrayMap(x -> demangle(addressToSymbol(x)), trace), '\n') AS trace_symbols,
    formatReadableQuantity(count(size) AS sz)
FROM system.trace_log
WHERE ((event_time >= start_time) AND (event_time <= end_time)) AND (trace_type = 'Real')
GROUP BY trace_symbols
ORDER BY sz DESC
LIMIT 10
SETTINGS allow_introspection_functions = 1

Query id: d4044db9-f065-4af7-be1f-99cbe963ac6c

Row 1:
──────
trace_symbols:                       DB::(anonymous namespace)::writeTraceInfo(DB::TraceType, int, siginfo_t*, void*)



std::__1::mutex::lock()
void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)
void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)
DB::MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipIndexes(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>>&&, std::__1::vector<std::__1::shared_ptr<DB::AlterConversions const>, std::__1::allocator<std::__1::shared_ptr<DB::AlterConversions const>>>&&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const>, std::__1::shared_ptr<DB::Context const> const&, DB::KeyCondition const&, std::__1::optional<DB::KeyCondition> const&, DB::UsefulSkipIndexes const&, DB::MergeTreeReaderSettings const&, std::__1::shared_ptr<Poco::Logger>, unsigned long, std::__1::vector<DB::ReadFromMergeTree::IndexStat, std::__1::allocator<DB::ReadFromMergeTree::IndexStat>>&, bool, bool)
DB::ReadFromMergeTree::selectRangesToRead(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>>, std::__1::vector<std::__1::shared_ptr<DB::AlterConversions const>, std::__1::allocator<std::__1::shared_ptr<DB::AlterConversions const>>>, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::SelectQueryInfo const&, std::__1::shared_ptr<DB::Context const>, unsigned long, std::__1::shared_ptr<std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, long, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const, long>>>>, DB::MergeTreeData const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::shared_ptr<Poco::Logger>, std::__1::optional<DB::ReadFromMergeTree::Indexes>&, bool)
DB::ReadFromMergeTree::selectRangesToRead(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>>, std::__1::vector<std::__1::shared_ptr<DB::AlterConversions const>, std::__1::allocator<std::__1::shared_ptr<DB::AlterConversions const>>>, bool) const
DB::ReadFromMergeTree::selectRangesToRead(bool) const
DB::ReadFromMergeTree::initializePipeline(DB::QueryPipelineBuilder&, DB::BuildQueryPipelineSettings const&)
DB::ISourceStep::updatePipeline(std::__1::vector<std::__1::unique_ptr<DB::QueryPipelineBuilder, std::__1::default_delete<DB::QueryPipelineBuilder>>, std::__1::allocator<std::__1::unique_ptr<DB::QueryPipelineBuilder, std::__1::default_delete<DB::QueryPipelineBuilder>>>>, DB::BuildQueryPipelineSettings const&)
DB::QueryPlan::buildQueryPipeline(DB::QueryPlanOptimizationSettings const&, DB::BuildQueryPipelineSettings const&)
DB::IInterpreterUnionOrSelectQuery::buildQueryPipeline(DB::QueryPlan&)
DB::IInterpreterUnionOrSelectQuery::execute()
DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*)
DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum)
DB::TCPHandler::runImpl()
DB::TCPHandler::run()
Poco::Net::TCPServerConnection::start()
Poco::Net::TCPServerDispatcher::run()
Poco::PooledThread::run()
Poco::ThreadImpl::runnableEntry(void*)


formatReadableQuantity(count(size)): 243.30 thousand

Row 2:
──────
trace_symbols:                       DB::(anonymous namespace)::writeTraceInfo(DB::TraceType, int, siginfo_t*, void*)



std::__1::mutex::lock()
void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)
DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)
DB::PullingAsyncPipelineExecutor::pull(DB::Block&, unsigned long)
DB::TCPHandler::runImpl()
DB::TCPHandler::run()
Poco::Net::TCPServerConnection::start()
Poco::Net::TCPServerDispatcher::run()
Poco::PooledThread::run()
Poco::ThreadImpl::runnableEntry(void*)


formatReadableQuantity(count(size)): 113.86 thousand
```

**Comment by canhld94** [comment]
cc @nickitat wdyt? I see you introduces this few months ago.

File: src/Interpreters/ConcurrentHashJoin.cpp [hunk 1]
```diff
@@ -23,6 +23,7 @@
 #include <Common/scope_guard_safe.h>
 #include <Common/setThreadName.h>
 #include <Common/typeid_cast.h>
+#include <Core/Settings.h>
```
Comment (nickitat) [inline src/Interpreters/ConcurrentHashJoin.cpp:26]: not needed

**Comment by nickitat** [comment]
> It's more likely that: we have many ConcurrentHashJoin and each of them using a pool with size of max_threads

pls adjust PR's title and change log entry to be more in sync with this

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit d089a5627280181a746d9183d28b1705906ac9d2 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69406/d089a5627280181a746d9183d28b1705906ac9d2/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/2b2cd8e3e076827fa61f98c9d422e76acb81e864/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Interpreters/ConcurrentHashJoin.cpp (hunk 1 above)
Comment (nickitat) [reply src/Interpreters/ConcurrentHashJoin.cpp:26]: all checks are green - let's dismiss it

**Review by nickitat — APPROVED** [review]

**Pull request merged**